python/ray/rllib/dqn/dqn.py

```diff
@@ -184,6 +207,69 @@ def get_gradient(self, cur_timestep):
             obses_t, actions, rewards, obses_tp1, dones = \
                 self.replay_buffer.sample(self.config["train_batch_size"])
             batch_idxes = None
+        replay_buffer_read_time = (time.time() - dt)
+        dt = time.time()
+        tuples_per_device = self.dqn_graph.multi_gpu_optimizer.load_data(
+            self.sess,
+            [obses_t, actions, rewards, obses_tp1, dones,
+             np.ones_like(rewards)])
+        per_device_batch_size = (
+            self.dqn_graph.multi_gpu_optimizer.per_device_batch_size)
+        num_batches = (int(tuples_per_device) // int(per_device_batch_size))
+        data_load_time = (time.time() - dt)
+        dt = time.time()
+        for _ in range(self.config["num_sgd_iter"]):
+            batches = list(range(num_batches))
+            random.shuffle(batches)
+            for i in batches:
+                self.dqn_graph.multi_gpu_optimizer.optimize(
+                    self.sess, i * per_device_batch_size)
+        sgd_time = (time.time() - dt)
+        dt = time.time()
+        if self.config["prioritized_replay"]:
+            dt = time.time()
+            td_errors = self.dqn_graph.compute_td_error(
+                self.sess, obses_t, actions, rewards, obses_tp1, dones,
+                np.ones_like(rewards))
+            dt = time.time()
+            new_priorities = (
+                np.abs(td_errors) + self.config["prioritized_replay_eps"])
+            self.replay_buffer.update_priorities(
+                batch_idxes, new_priorities)
+        prioritization_time = (time.time() - dt)
```

Comment:
are we sure we want to keep this timing code? if you really want, I would put everything into a `info` dict and return that

Reply:
Yes, timing is important part for a experimental feature. I changed it to a dict.

I was also debating leaving out the multi-gpu part, but it's probably better to leave it in otherwise that stuff tends to code rot.